packages/element/src/create-interpolate-element.js

```diff
@@ -0,0 +1,235 @@
+
+	/**
+	 * This short circuits the process if the conversion map has an invalid config.
+	 */
+	if ( ! searchString || ! conversionConfig ) {
+		return potentialElement;
+	}
+
+	const match = getMatchFromString(
+		potentialElement,
+		searchString,
+		conversionConfig
+	);
+
+	// if there is no match for this string, then that means it is not an element
+	// so just return as is to be used as a direct child.
+	if ( match === null ) {
+		return potentialElement;
+	}
+
+	// if the full match returned equals the potential element, then we know
+	// we can create the element and restart the conversion on any children if
+	// necessary.
+	if ( match[ 0 ] === potentialElement ) {
+		// remove this item from the conversion map because it's no longer needed.
+		conversionMap.shift();
+
+		if ( getHasPropValue( conversionConfig ) ) {
+			// if value is a react element, then need to wrap in Fragment with a key
+			// to prevent key warnings.
+			if ( isValidElement( conversionConfig.value ) ) {
+				keyIndex++;
+				return <Fragment key={ keyIndex }>{ conversionConfig.value }</Fragment>;
+			}
+			return conversionConfig.value;
+		}
+		keyIndex++;
+		return getHasChildren( conversionConfig ) ?
+			createElement(
+				conversionConfig.tag,
+				{ ...conversionConfig.props, key: keyIndex },
+				recursiveCreateElement( match[ 1 ], conversionMap )
+			) :
+			createElement(
+				conversionConfig.tag,
+				{ ...conversionConfig.props, key: keyIndex }
+			);
+	}
+
+	// still here, so we need to split on the full match and loop through each
+	return flatMap(
+		potentialElement.split( getSplitRegEx( match[ 0 ] ) )
+			.filter( ( value ) => !! value ),
+		( element ) => {
+			return recursiveCreateElement( element, conversionMap );
+		}
+	);
+};
+
+/**
+ * This function creates an interpolated element from a passed in string with
+ * specific tags matching how the string should be converted to an element via
+ * the conversion map value.
+ *
+ * @example
+ * For example, for the given string:
+ *
+ * "This is a <span%1>string</span%1> with <a%1>a link</a%1>, a self-closing
+ * %1$s tag and a plain value %2$s"
+ *
+ * You would have something like this as the conversionMap value:
+ *
+ * ```js
+ * {
+ *     'span%1': { tag: CustomComponent, props: {} },
+ *     'a%1': { tag: 'a', props: { href: 'https://github.com' } },
+ *     '%1$s': { tag: CustomComponentB, props: {} },
+ *     '%2$s': { value: 'custom value' },
+ * }
+ * ```
+ *
+ * @param {string}  interpolatedString  The interpolation string to be parsed.
+ * @param {Object} conversionMap        The map used to convert the string to
+ *                                      a react element.
+ *
+ * @return {Element}  A react element.
+ */
+const createInterpolateElement = ( interpolatedString, conversionMap ) => {
+	keyIndex = -1;
+	return createElement(
+		Fragment,
+		{},
+		recursiveCreateElement(
+			interpolatedString,
+			Object.entries( conversionMap )
```

Comment:
I don't know if this has been discussed elsewhere, but I have doubts about an interface that is by nature unordered (an object) which is then consumed sequentially (see instances of `conversionMap.shift()` _et al._).

As an outside developer, I would expect one of two scenarios:

1. `conversionMap` is an object keyed on search strings, in which case the order of the object's values is irrelevant.
2. `conversionMap` is an array of conversion configs, in which case their order needs to follow the interpolated string. A variant of this would be a `printf`-style argument list of `( interpolateString, ...conversionConfigs )`:

```js
createInterpolateElement(
	'This is a complex string having a %1$s value, with a <a1>nested <em1>%2$s</em1> link</a1> and value: %3$s',
	{ value: 'concrete' },
	{ tag: 'a', props: {} },
	{ tag: 'em', props: {} },
	{ value: 'value' },
	{ value: <TestComponent /> },
)
```

The other thing on my mind is whether we should allow numerical tags. Because, if so, one could reasonably expect the equivalence:

```js
createInterpolateElement(
    'Hello, %2$s. You have %1$s new message(s).',
    'User',
    42
) === createElement( Fragment, {}, [
    'Hello, ',
    42,
    '. You have ',
    'User',
    ' new message(s).'
] )
```

And I do apologise for raising questions on the interface itself at this stage of development!

Reply:
> And I do apologise for raising questions on the interface itself at this stage of development!

No apologies needed!  I'm happy to get feedback on anything in this pull :)

> I don't know if this has been discussed elsewhere, but I have doubts about an interface that is by nature unordered (an object) which is then consumed sequentially (see instances of conversionMap.shift() et al.).

Ya, I agree.  I think I reacted to quickly to some feedback @gziolo gave (which doesn't exist now I think because of a rebase?) about the [original shape](https://github.com/WordPress/gutenberg/pull/16374) which was something like this:

```js
		[
			[ 'span%1', { tag: CustomComponent, props: {}, hasChildren: true } ],
			[ 'a%1', { tag: 'a', props: { href: url }, hasChildren: true } ],
			[ '%1$s', { tag: CustomComponentB, props: {}, hasChildren: false } ],
			[ '%2$s', { value: 'custom value' } ],
		]
```

His feedback was to do the object shape like what I have (which I think was great) primarily to promote the necessity for tag identifiers for each element in the map to be unique. I thought I could get away with preserving the existing logic by just converting things to an array,  but I totally get what you're saying about the current shape implying that order doesn't matter, when in fact, it does.

I think it's important we preserve the ability to name tags explicitly both to clearly differentiate between them, and also to potentially add the ability to add more context for translators (cc @akirk on this... I think he might be able to elaborate on that point).

So as far as interface, assuming we preserve the existing logic (which imo is a fairly simple straightforward parse avoiding some of the typical problems with nested tags), the following requirements exist on the map:

1. identifiers must me unique.
2. The map must be in order of tags in the interpolated string.

Additional requirements that I have picked up from other sources are:

1. Context must be preserved for translators (i.e. identify the value is a link).  _Only_ tokens is not acceptable.
2. Having the ability to do not only `<a1>` but also `<a link to marketing>` to provide context would be useful.

So with that in mind, how do you feel about an interface that is something like this:

```jsx
__experimentalCreateInterpolateElement(
  'This is a complex string having a <concrete> value, with ' +
	'a <a1>nested <em1><value></em1> link</a1> and value: <A Special Component />,
  { value: 'concrete' },
  { tag: 'a', props: {} },
  { tag: 'em', props: {} },
  { value: 'value' },
  { custom: 'A Special Component' value: <TestComponent /> }
);
```

The following criteria exist for the arguments in this proposal:

1. They are in order according to how they appear in the string.
2. No need to provide a custom identifier as it will use the string representation of whatever the `tag` or `value` prop is.  However, they simply will be incremented internally for the matches (which means that the interpolation string must have numbers identifying unique tags.  We kind of need to keep this in place otherwise we can get into some pretty hairy nesting match expressions for the regex.
3. If one wants to be explicit for tag name in the string, they can by simply providing a `custom` prop with the name being used in the string.

Note, this proposal also gets rid of the numerical `%1$s` type tags that can get confused with `printf` behaviour. Thoughts?

cc @swissspidy or @ocean90 - I'd appreciate your thoughts on this interface as well from the perspective of both the string that translators would see, and the usability of this from the standpoint of code consuming the interface.

Reply:
> 1. `conversionMap` is an object keyed on search strings, in which case the order of the object's values is irrelevant.

I assumed this scenario based on the examples and tests provided.

Reply:
I wasn't looking at the whole picture here; in particular, where in all this would `__` be called. I assume that one should be able to construct an expression tree such as:

```js
createInterpolateElement(
	__( 'This is a <span%1>string</span%1> with <a%1>a link</a%1>, a self-closing %1$s component and a plain value: %2$s' ),
	{
		// conversionMap
	}
);
```

(the above example was taken from the description of #17376)

In other words, `createInterpolateElement` should accept a translated string as its first argument. In the course of translation — the sentence structure in the target locale being entirely dependent on the language — we must **assume that tokens have been reordered**.

With that in mind, only two high-level approaches are possible:

- We go "full `sprintf`" and require all tokens to be of the `%1$s` kind, so that then the interface can be order-sensitive: `createInterpolateElement( myString, config1, config2, … )`
- We go with unique slugs as tokens. This could look like what we have right now (`createInterpolateElement( testString, { em1: { tag: 'em', props: {} }, a1: { tag: 'a', props: {} } } )` but without imposing any particular order on the conversion map. If I may add: since this would imply a requirement of slug unicity, then a proper map (JS object) would definitely be the way to go, because then the data type itself would enforce the requirement (in contrast, a pairs format like `[ [ key1, value1 ], [ key2, value2] ]` would require extra logic to check the unicity of the keys).

Whatever the decisions for the interface, the implementation itself will be order-insensitive: internally it must be able to access the conversion configurations non-sequentially.

* * *

> I think it's important we preserve the ability to name tags explicitly both to clearly differentiate between them, and also to potentially add the ability to add more context for translators

IMO slugs could help avoid human error, both for the developer and the translator, but would do little for context. To provide good context, the usual methods would still apply (`translators:` string and `_x()`-type functions). The slugs would still have to be explained, so the gain is small. :)

Reply:
> we must assume that tokens have been reordered.

Ya I concur.  It does complicate things (especially for nesting) but it's pretty clear it's a requirement.  I'll work on implementing that the next chance I get to work on this pull.

Reply:
> IMO slugs could help avoid human error, both for the developer and the translator, but would do little for context. To provide good context, the usual methods would still apply (translators: string and _x()-type functions). The slugs would still have to be explained, so the gain is small. :)

I agree here too, I wasn't suggesting that they'd be used in lieu of `_x()` type functions but more as a compliment. It's something that was brought up by @akirk in conversations I had with him so I mostly recorded it here for context.  I think descriptive slugs also have the potential to make things _more_ confusing for translators because they might see that as something to translate itself!  So while we can support it, I think best practices would be to limit custom tags to variations of common elements (`<a>`, `<a1>`, `<span>`, `<strong>`, `<span>` etc).

Reply:
Descriptive placeholders are good for translators. Translation software will protect placeholders and requires them to appear in the translation.

Reply:
see https://github.com/WordPress/gutenberg/pull/17376#issuecomment-541433887